File: apps/admin-v2/src/domain/entities/reservation.ts
export type ReservationStatus = 'pending' | 'confirmed' | 'checked_in' | 'in_progress' | 'completed' | 'cancelled' | 'no_show';
export type ReservationAction = 'confirm' | 'start' | 'complete' | 'cancel' | 'no_show';

export type ReservationItemType = 'service_variant' | 'product';

export type ReservationPaymentStatus = 'unpaid' | 'paid';
export type ReservationPaymentMethod = 'transfer' | 'card' | 'cash';
export type ReservationPaymentTiming =
  | 'prepay_required'
  | 'at_pickup'
  | 'at_completion'
  | 'flexible'
  | 'none';

export interface ReservationItem {
  id: string;
  reservationId: string;
  itemType: ReservationItemType;
  refId: string;
  /** Parent service for variant items. Empty for product lines. */
  serviceId?: string | null;
  label: string;
  qty: number;
  unitPrice: number;
  lineTotal: number;
  sortOrder: number;
  createdAt?: Date;
}

export interface ReservationItemChange {
  id: string;
  action: 'added' | 'removed' | 'upgraded' | 'downgraded' | 'price_override';
  itemType: string | null;
  label: string | null;
  oldPrice: number | null;
  newPrice: number | null;
  reason: string | null;
  changedBy: { id: string; name: string } | null;
  changedAt: Date;
}

export interface BillingSnapshot {
  docType: 'ruc' | 'cedula' | 'passport' | 'final_consumer';
  docNumber: string;
  legalName: string;
  email: string | null;
  address: string | null;
  phone: string | null;
  source: 'profile' | 'manual';
  capturedAt: string;
}

export interface ReservationClientResource {
  label: string | null;
  data: Record<string, unknown> | null;
  plate: string | null;
  brand: string | null;
  model: string | null;
  color: string | null;
  type?: string | null;
}

export interface ReservationService {
  name: string;
  price: string;
}

export interface ReservationClient {
  name: string;
  email: string;
}

export interface Reservation {
  id: string;
  clientId: string;
  clientResourceId: string;
  businessResourceId: string | null;
  serviceId: string;
  serviceVariantId: string | null;
  assignedTo: string | null;
  scheduledAt: Date;
  estimatedEnd: Date;
  status: ReservationStatus;
  notes: string | null;
  cancelledAt: Date | null;
  cancelReason: string | null;
  createdBy: string;
  createdAt: Date;
  checkedInAt: Date | null;
  billingSnapshot: BillingSnapshot | null;
  paymentStatus: ReservationPaymentStatus;
  paymentMethod: ReservationPaymentMethod | null;
  paidAt: Date | null;
  paymentReference: string | null;
  /** Bank slug (`pichincha`, `pacifico`…) when paid via transfer. Null
      otherwise. Free-form string so tenants can add regional banks. */
  paymentBank: string | null;
  invoiced: boolean;
  invoicedAt: Date | null;
  invoiceExternalId: string | null;
  invoiceStatus: string | null;
  invoiceClaveAcceso: string | null;
  invoiceNumeroAutorizacion: string | null;
  invoiceError: string | null;
  clientResource?: ReservationClientResource;
  service?: ReservationService;
  client?: ReservationClient;
  /** Compact roll-up of every item label so list views can render
      "Lavada + Aspirado +1 más" without firing a /items request per
      row. Backend ships it when `items` is eager-loaded. */
  servicesSummary?: { count: number; labels: string[] };
}

export interface AvailableSlot {
  start: Date;
  end: Date;
  available: number;
}

export interface ReservationFilters {
  dateFrom?: string;
  dateTo?: string;
  status?: ReservationStatus;
  serviceId?: string;
  page?: number;
}

File: apps/admin_mobile/lib/domain/entities/reservation.dart
import 'package:equatable/equatable.dart';

enum ReservationStatus {
  pending,
  confirmed,
  inProgress,
  completed,
  cancelled,
  noShow;

  String get apiValue {
    switch (this) {
      case ReservationStatus.pending:
        return 'pending';
      case ReservationStatus.confirmed:
        return 'confirmed';
      case ReservationStatus.inProgress:
        return 'in_progress';
      case ReservationStatus.completed:
        return 'completed';
      case ReservationStatus.cancelled:
        return 'cancelled';
      case ReservationStatus.noShow:
        return 'no_show';
    }
  }

  static ReservationStatus fromApi(String value) {
    switch (value) {
      case 'pending':
        return ReservationStatus.pending;
      case 'confirmed':
        return ReservationStatus.confirmed;
      case 'in_progress':
        return ReservationStatus.inProgress;
      case 'completed':
        return ReservationStatus.completed;
      case 'cancelled':
        return ReservationStatus.cancelled;
      case 'no_show':
        return ReservationStatus.noShow;
      default:
        throw ArgumentError('Unknown ReservationStatus: $value');
    }
  }
}

enum ReservationAction {
  confirm,
  start,
  complete,
  cancel;

  String get apiValue {
    switch (this) {
      case ReservationAction.confirm:
        return 'confirm';
      case ReservationAction.start:
        return 'start';
      case ReservationAction.complete:
        return 'complete';
      case ReservationAction.cancel:
        return 'cancel';
    }
  }
}

class Reservation extends Equatable {
  final int id;
  final int clientId;
  final int clientResourceId;
  final int serviceId;
  final int? assignedTo;
  final DateTime scheduledAt;
  final DateTime estimatedEnd;
  final ReservationStatus status;
  final String? notes;
  final DateTime? cancelledAt;
  final String? cancelReason;
  final int createdBy;
  final DateTime createdAt;
  final String? clientResourceLabel;
  final String? clientResourcePlate;
  final String? serviceName;
  final double? servicePrice;
  final String? clientName;
  final String? clientEmail;

  const Reservation({
    required this.id,
    required this.clientId,
    required this.clientResourceId,
    required this.serviceId,
    this.assignedTo,
    required this.scheduledAt,
    required this.estimatedEnd,
    required this.status,
    this.notes,
    this.cancelledAt,
    this.cancelReason,
    required this.createdBy,
    required this.createdAt,
    this.clientResourceLabel,
    this.clientResourcePlate,
    this.serviceName,
    this.servicePrice,
    this.clientName,
    this.clientEmail,
  });

  @override
  List<Object?> get props => [id];
}

class AvailableSlot extends Equatable {
  final DateTime start;
  final DateTime end;
  final int available;

  const AvailableSlot({
    required this.start,
    required this.end,
    required this.available,
  });

  @override
  List<Object?> get props => [start, end, available];
}

class ReservationFilters extends Equatable {
  final DateTime? dateFrom;
  final DateTime? dateTo;
  final ReservationStatus? status;
  final int? serviceId;
  final int? page;

  const ReservationFilters({
    this.dateFrom,
    this.dateTo,
    this.status,
    this.serviceId,
    this.page,
  });

  @override
  List<Object?> get props => [dateFrom, dateTo, status, serviceId, page];
}
